### PR TITLE
feat: line item adjustments

### LINKED ISSUE #1229: Introduce LineItemAdjustment
We would like to introduce a notion of adjustments on line items like so:
```
LineItem
id
unit_price
adjustments: [
  amount: -100,
  description: "discount" 
]
```

This change will require us to refactor the TotalsService, such that these adjustments are reflected in calculations. Something along the lines of:
```
getDiscountTotal(items) {
   return items.reduce((acc, next) => {   return acc + next.quantity * unit_price + sum(adjustments.amount) + taxes }, 0)
}
```

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- adds support for line item adjustments #1229 

### Why

- ease totals calculation
- be able to support a use case where store operators want to add custom adjustments to an order (not implemented in this PR)

### How

- Add a `LineItemAdjustment` entity to line items (#1161 + #1241)
- Adjust `TotalsService` to use the line item adjustments when calculating discount total (#1167)
- Adjust `LineItem.generate` to use line item adjustments when generating line items (#1243)
- Adjust `LineItem.createReturnLines` to use line item adjustments on return lines (#1242)
- Adjust `CartService` to create/delete/update line item adjustments on various cart business events (#1242)
- Adjust `SwapService.createCart` to create line item adjustments on `additional_items` when creating a swap cart
- Refactored `CartService.applyDiscount` method (#1242)

### Testing

- Unit tests for line item adjustment service methods (#1161 + #1242)
- Adjust unit tests for carts
- Adjust unit tests for swaps
- Adjust unit tests for discount
- Adjust unit tests for totals
- Integration tests for ensuring the db constraints work as expected (#1279)
- Integration tests for ensuring line item adjustments are created on various cart business events (#1242)
- Integration tests for ensuring item adjustments are created on return items

closes #1229 #1168

### CODE REVIEW

**Review by srindom — COMMENTED** [review]
LGTM! 🚀

**Review by srindom — APPROVED** [review]

**Review by adrien2p — COMMENTED** [review]
@srindom I know that the PR has already been merged but there is some mistakes that will be fixed by #1306.
I have only reviewed the cart service since my PR was about that service :)

File: packages/medusa/src/services/line-item-adjustment.ts [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+      // then do nothing
+      if (
+        !generatedLineItem.allow_discounts ||
+        generatedLineItem.is_return ||
+        !cart?.discounts?.length
+      ) {
+        return []
+      }
+
+      const [discount] = cart.discounts.filter(
```
Comment (adrien2p) [inline packages/medusa/src/services/line-item-adjustment.ts:208]: @srindom just for my knowledge, only one discount can appears at the time?

Comment (adrien2p) [inline packages/medusa/src/services/line-item-adjustment.ts:208]: @srindom just for my knowledge, only one discount can appears at the time?

**Pull request merged**